### PR TITLE
feat(metaclient(dataflow)): add etcd kv client implement

### PR DESCRIPTION
**Pull request by maxshuang**
close issue #157

### CODE REVIEW

File: pkg/metaclient/kvclient/etcdkv/etcd_client.go [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+type etcdImpl struct {
+	cli           *clientv3.Client
+	compactCancel context.CancelFunc
+	revCh         chan int64
+	latestRevT    int64
+	closeMu       sync.Mutex
+	wg            *sync.WaitGroup
+}
+
+func startCompactRoutine(ctx context.Context, wg *sync.WaitGroup, ch <-chan int64, cli *clientv3.Client) {
```
Comment (liuzix) [inline pkg/metaclient/kvclient/etcdkv/etcd_client.go:46]: Please add a comment justifying this function.

File: pkg/metaclient/kvclient/etcdkv/etcd_client.go [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+		revCh:         revCh,
+		latestRevT:    time.Now().Unix(),
+		wg:            &wg,
+	}
+	return c, nil
+}
+
+func (c *etcdImpl) getLatestRev(h *pb.ResponseHeader) {
+	now := time.Now().Unix()
+	if (now - atomic.LoadInt64(&c.latestRevT)) > 1 {
```
Comment (liuzix) [inline pkg/metaclient/kvclient/etcdkv/etcd_client.go:111]: Please clarify the magic number `1`.

File: pkg/metaclient/kvclient/etcdkv/etcd_client.go [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+		atomic.StoreInt64(&c.latestRevT, now)
+		c.revCh <- h.Revision
```
Comment (liuzix) [inline pkg/metaclient/kvclient/etcdkv/etcd_client.go:113]: Could `c.revCh` be full here?

File: pkg/metaclient/kvclient/etcdkv/etcd_client.go [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+func (c *etcdImpl) getEtcdOptions(op metaclient.Op) []clientv3.OpOption {
+	etcdOps := make([]clientv3.OpOption, 0, 3)
+	switch {
+	case op.IsOptsWithPrefix():
+		etcdOps = append(etcdOps, clientv3.WithPrefix())
+	case op.IsOptsWithFromKey():
+		etcdOps = append(etcdOps, clientv3.WithFromKey())
+	case op.IsOptsWithRange():
+		etcdOps = append(etcdOps, clientv3.WithRange(string(op.RangeBytes())))
+	}
```
Comment (liuzix) [inline pkg/metaclient/kvclient/etcdkv/etcd_client.go:126]: Do we need a `default` branch?

File: pkg/metaclient/kvclient/etcdkv/etcd_client.go [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+	case op.IsTxn():
+		ops := op.Txn()
+		etcdOps := make([]clientv3.Op, 0, len(ops))
+		for _, sop := range ops {
+			etcdOps = append(etcdOps, c.getEtcdOp(sop))
+		}
+		return clientv3.OpTxn(nil, etcdOps, nil)
+	}
+
+	panic("unknown op type")
```
Comment (liuzix) [inline pkg/metaclient/kvclient/etcdkv/etcd_client.go:149]: Can we move this to `default`?

File: pkg/metaclient/kvclient/etcdkv/etcd_client.go [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+		cli:           cli,
+		compactCancel: cancel,
+		revCh:         revCh,
+		latestRevT:    time.Now().Unix(),
+		wg:            &wg,
+	}
+	return c, nil
+}
+
+func (c *etcdImpl) getLatestRev(h *pb.ResponseHeader) {
```
Comment (liuzix) [inline pkg/metaclient/kvclient/etcdkv/etcd_client.go:109]: Please comment on the purpose and usage of this function.

File: pkg/metaclient/kvclient/etcdkv/etcd_client.go [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+	}
+
+	panic("unknown op type")
+}
+
+func (c *etcdImpl) Put(ctx context.Context, key, val string) (*metaclient.PutResponse, error) {
+	op := metaclient.OpPut(key, val)
+	etcdResp, err := c.cli.Do(ctx, c.getEtcdOp(op))
+	if err != nil {
+		return nil, cerrors.ErrMetaOpFail.FastGenByArgs(err)
```
Comment (liuzix) [inline pkg/metaclient/kvclient/etcdkv/etcd_client.go:156]: Please justify your choice of using `FastGenByArgs`

File: pkg/metaclient/kvclient/etcdkv/etcd_client.go [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+		txnRsp := makeTxnResp(rsp)
+		return txnRsp.OpResponse(), nil
+	default:
+		panic("Unknown op")
+	}
+}
+
+type etcdTxn struct {
+	clientv3.Txn
+	mu  sync.Mutex
```
Comment (liuzix) [inline pkg/metaclient/kvclient/etcdkv/etcd_client.go:234]: Please make it clear what fields `mu` is intended to protect.

File: pkg/metaclient/kvclient/etcdkv/etcd_client.go [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+	}
+
+	t.ops = append(t.ops, etcdOps...)
+	return t
+}
+
+func (t *etcdTxn) Commit() (*metaclient.TxnResponse, error) {
+	t.mu.Lock()
+	if t.Err != nil {
+		return nil, t.Err
```
Comment (liuzix) [inline pkg/metaclient/kvclient/etcdkv/etcd_client.go:290]: `t.mu.Unlock()`

File: pkg/metaclient/kvclient/etcdkv/etcd_client.go [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+	}
+	if t.committed {
+		t.Err = cerrors.ErrMetaCommittedTxn
+		return nil, t.Err
```
Comment (liuzix) [inline pkg/metaclient/kvclient/etcdkv/etcd_client.go:294]: `t.mu.Unlock()`

File: pkg/metaclient/kvclient/etcdkv/util.go [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+	}
+
+	return resp
+}
+
+func makeGetResp(etcdResp *clientv3.GetResponse) *metaclient.GetResponse {
+	kvs := make([]*metaclient.KeyValue, 0, len(etcdResp.Kvs))
+	for _, kv := range etcdResp.Kvs {
+		if kv.Version == 0 {
+			// This key has been deleted, don't return to user
```
Comment (liuzix) [inline pkg/metaclient/kvclient/etcdkv/util.go:26]: When can this happen?

File: pkg/metaclient/op.go [hunk 12]
```diff
@@ -52,7 +52,7 @@ type Op struct {
 
 var EmptyOp Op = Op{}
 
-// accessors
+// accessors/mutors
```
Comment (liuzix) [inline pkg/metaclient/op.go:55]: Do you mean `mutators`?

File: pkg/metaclient/kvclient/etcdkv/etcd_client.go (hunk 8 above)
Comment (maxshuang) [reply pkg/metaclient/kvclient/etcdkv/etcd_client.go:234]: Almost all inner members, so I don't specify the fields.

File: pkg/metaclient/client.go [hunk 13]
```diff
@@ -6,7 +6,11 @@ const (
 	TypeKVClient ClientType = iota /* KV client style，like etcd/consul/TiKV/redis or even SQL backend*/
 )
 
+type Client interface {
```
Comment (amyangfei) [inline pkg/metaclient/client.go:9]: `Closer` is better.

File: pkg/metaclient/kvclient/etcdkv/etcd_client.go (hunk 4 above)
Comment (maxshuang) [reply pkg/metaclient/kvclient/etcdkv/etcd_client.go:68]: Without a `default` branch seems ok here.

File: pkg/metaclient/kvclient/etcdkv/util.go (hunk 11 above)
Comment (maxshuang) [reply pkg/metaclient/kvclient/etcdkv/util.go:26]: I find the comment on kv.version with `A delete resets the version to zero`.
I will check it on the server side later.

File: pkg/metaclient/kvclient/etcdkv/etcd_client_test.go [hunk 14]
```diff
@@ -0,0 +1,310 @@
...
+
+	grsp, gerr := cli.Get(ctx, "hello", metaclient.WithRange("s"))
+	require.Nil(t, gerr)
+	require.Len(t, grsp.Kvs, 3) // hello1, hello2, interesting
+	require.Equal(t, string(grsp.Kvs[0].Key), "hello1")
+	require.Equal(t, string(grsp.Kvs[0].Value), "world1")
+	require.Equal(t, string(grsp.Kvs[1].Key), "hello2")
+	require.Equal(t, string(grsp.Kvs[1].Value), "world2")
+	require.Equal(t, string(grsp.Kvs[2].Key), "interesting")
+	require.Equal(t, string(grsp.Kvs[2].Value), "world")
```
Comment (amyangfei) [inline pkg/metaclient/kvclient/etcdkv/etcd_client_test.go:161]: Better to make the test case into table format, such as
```go
type kv struct {
    key string
    value string
}

input = []struct{
    kv kv
} {
    {"key1", "value1"},
    {"key2", "value2"},
}

query := []struct {
    key string    
    opts Op
    expected []kv 
} {
    ....
}
```
The case will be more clear and easy to maintain.

**Review by amyangfei — APPROVED** [review]

**Pull request merged**